src/fields/python_field_base.hh

```diff
@@ -0,0 +1,201 @@
+
+/// Helper class, holds data of one field
+class FieldCacheProxy
+{
+public:
+    /// Constructor
+    FieldCacheProxy(std::string field_name, std::vector<ssize_t> shape, std::vector<double> field_cache_ptr)
+    : field_name_(field_name), shape_(shape), field_cache_ptr_(field_cache_ptr)
+    {
+        //ASSERT_EQ(shape.size(), 2);
+    }
+
+    /// Getters
+    const std::string &field_name() const { return field_name_; }
+    ssize_t n_rows() const { return shape_[0]; }
+    ssize_t n_cols() const { return shape_[1]; }
+    std::vector<double> &field_cache_ptr() { return field_cache_ptr_; }
+private:
+    std::string field_name_;
+    std::vector<ssize_t> shape_;
+    std::vector<double> field_cache_ptr_;
+};
+
+
+class PythonFieldBase
+{
+protected:
+    py::array create_array_with_data(std::vector<double> &data, ssize_t n_rows, ssize_t n_cols)
+    {
+        ssize_t              size    = data.size() / (n_rows*n_cols);
+        ssize_t              ndim    = 2;
+        std::vector<ssize_t> shape   = { n_rows*n_cols , size };
+        std::vector<ssize_t> strides = { (long int)(sizeof(double)*size) , sizeof(double) };
+
+        // create 2-D NumPy array
+        return  py::array(py::buffer_info(
+            &data[0],                                /* data as contiguous array  */
+            sizeof(double),                          /* size of one scalar        */
+            py::format_descriptor<double>::format(), /* data type                 */
+            ndim,                                    /* number of dimensions      */
+            shape,                                   /* shape of the matrix       */
+            strides                                  /* strides for each axis     */
+        ));
+    }
+
+    py::array create_array(ssize_t n_rows, ssize_t n_cols, ssize_t size, bool fill = true)
+    {
+        std::vector<double> result(size*n_rows*n_cols, 0.0);
+        if (fill) {
+            for (uint i=0; i<result.size(); ++i) result[i] = 1.0 + i;
+        }
+
+        ssize_t              ndim    = 2;
+        std::vector<ssize_t> shape   = { n_rows*n_cols , size };
+        std::vector<ssize_t> strides = { (long int)(sizeof(double)*size) , sizeof(double) };
+
+        // create 2-D NumPy array
+        return  py::array(py::buffer_info(
+            result.data(),                           /* data as contiguous array  */
+            sizeof(double),                          /* size of one scalar        */
+            py::format_descriptor<double>::format(), /* data type                 */
+            ndim,                                    /* number of dimensions      */
+            shape,                                   /* shape of the matrix       */
+            strides                                  /* strides for each axis     */
+        ));
+    }
+public:
+    PythonFieldBase()
+    {}
+
+    PythonFieldBase(std::vector<FieldCacheProxy> &data, FieldCacheProxy &result)
```

Comment:
The constructor should match the constructor of PythonFieldBase (in Python).
In fact we may possibly completely get rid of this class and just call the Python code, no need to call the C++ back.:
1. C++ calls PythonFieldBase.create class method passing in the user class. 
  This creates the unique instance of the user class and store it in a dictionary  as a class attribute of PythonFieldBase. The instance is also returned to C++. Futher referred as `instance`.
2. C++ calls `instance._used_fields` to get used fiedls
3. C++ calls `instance._cache_reinit(t:float, inputs:Dict[str, CacheProxy], result: CacheProxy)`, the instance merge all inputs and check that all used fields are passed in. As the instance is just one for the single user class various FieldPython fields may call `_cache_reinit` with different arguments.
All results are stored in a single dictionary of the output fields.
4. C++ calls `instance.call(field_name)` implemented in PythonFieldBase during cache_update. That calls method `<field_name>` of the user class and store the returned value to the result array.